x/stake/keeper/val_state_change.go

```diff
@@ -0,0 +1,287 @@
+		newPowerBytes := validator.ABCIValidatorPowerBytes(k.cdc)
+
+		// update the validator set if power has changed
+		if !ok || !bytes.Equal(powerBytes, newPowerBytes) {
+			updates = append(updates, validator.ABCIValidatorUpdate())
+		}
+
+		// validator still in the validator set
+		delete(last, opbytes)
+
+		// set the bonded validator index
+		store.Set(GetBondedValidatorIndexKey(operator), newPowerBytes)
+
+		// keep count
+		count++
+
+	}
+
+	// sort the map keys for determinism
+	noLongerBonded := make([][]byte, len(last))
+	index := 0
+	for oper := range last {
+		operator := make([]byte, sdk.AddrLen)
+		copy(operator[:], oper[:])
+		noLongerBonded[index] = operator
+		index++
+	}
+	sort.SliceStable(noLongerBonded, func(i, j int) bool {
+		return bytes.Compare(noLongerBonded[i], noLongerBonded[j]) == -1
+	})
+
+	// iterate through the sorted no-longer-bonded validators
+	// any validators left in `last` are no longer bonded
+	for _, operator := range noLongerBonded {
+		// fetch the validator
+		validator := k.mustGetValidator(ctx, sdk.ValAddress(operator))
+
+		// bonded to unbonding
+		k.bondedToUnbonding(ctx, validator)
+
+		// remove validator if it has no more tokens
+		if validator.Tokens.IsZero() {
+			k.RemoveValidator(ctx, validator.OperatorAddr)
+		}
+
+		// delete from the bonded validator index
+		store.Delete(GetBondedValidatorIndexKey(operator))
+
+		// update the validator
+		updates = append(updates, validator.ABCIValidatorUpdateZero())
+	}
+
+	return updates
+}
+
+// Validator state transitions
+
+func (k Keeper) bondedToUnbonding(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Bonded {
+		panic(fmt.Sprintf("bad state transition bondedToUnbonding, validator: %v\n", validator))
+	}
+	return k.beginUnbondingValidator(ctx, validator)
+}
+
+func (k Keeper) unbondingToBonded(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Unbonding {
+		panic(fmt.Sprintf("bad state transition unbondingToBonded, validator: %v\n", validator))
+	}
+	return k.bondValidator(ctx, validator)
+}
+
+func (k Keeper) unbondedToBonded(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Unbonded {
+		panic(fmt.Sprintf("bad state transition unbondedToBonded, validator: %v\n", validator))
+	}
+	return k.bondValidator(ctx, validator)
+}
+
+func (k Keeper) unbondingToUnbonded(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Unbonded {
+		panic(fmt.Sprintf("bad state transition unbondingToBonded, validator: %v\n", validator))
+	}
+	return k.completeUnbondingValidator(ctx, validator)
+}
+
+// send a validator to jail
+func (k Keeper) JailValidator(ctx sdk.Context, validator types.Validator) {
+	if validator.Jailed {
+		panic(fmt.Sprintf("cannot jail already jailed validator, validator: %v\n", validator))
+	}
+
+	pool := k.GetPool(ctx)
+	k.DeleteValidatorByPowerIndex(ctx, validator, pool)
```

Comment:
Silly question, but in these state transition methods, why do we need to `DeleteValidatorByPowerIndex` and then set it again after we update the validator? Won't the set override existing values?

Reply:
No - we have to delete the old value when we know the old validator power.